include/mbedtls/mps/reader.h

```diff
@@ -0,0 +1,774 @@
+ * New buffers can be requested through calls to mbedtls_reader_get(),
+ * while previously obtained input buffers can be marked processed
+ * through calls to mbedtls_reader_consume(), emptying the list of
+ * input buffers and invalidating them from the consumer's perspective.
+ * The consumer need not be aware of the distinction between consumer
+ * and producer mode, because he only interfaces with the reader
+ * when the latter is in consuming mode.
+ *
+ * From the perspective of the producer, the state of the reader
+ * is one of the following:
+ * - Attached: An incoming data buffer is currently
+ *             being managed by the reader, and
+ * - Unset: No incoming data buffer is currently
+ *          managed by the reader, and all previously
+ *          handed incoming data buffers have been
+ *          fully processed.
+ * - Accumulating: No incoming data buffer is currently
+ *                 managed by the reader, but some data
+ *                 from the previous incoming data buffer
+ *                 hasn't been processed yet and is internally
+ *                 held back.
+ * The Unset and Accumulating states belong to producing mode,
+ * while the Attached state belongs to consuming mode.
+ *
+ * Transitioning from Unset or Accumulating to Attached is
+ * done via calls to mbedtls_reader_feed(), while transitioning
+ * from Consuming to either Unset or Accumulating (depending
+ * on what has been processed) is done via mbedtls_reader_reclaim().
+ *
+ * The following diagram depicts the producer-state progression:
+ *
+ *        +------------------+             reclaim
+ *        |      Unset       +<-------------------------------------+       get
+ *        +--------|---------+                                      |   +------+
+ *                 |                                                |   |      |
+ *                 |                                                |   |      |
+ *                 |                feed                  +---------+---+--+   |
+ *                 +-------------------------------------->    Attached    <---+
+ *                                                        |       /        |
+ *                 +-------------------------------------->    Consuming   <---+
+ *                 |     feed, enough data available      +---------+---+--+   |
+ *                 |     to serve previous consumer request         |   |      |
+ *                 |                                                |   |      |
+ *        +--------+---------+                                      |   +------+
+ *   +---->   Accumulating   |<-------------------------------------+    commit
+ *   |    +---+--------------+      reclaim, previous read request
+ *   |        |                        couldn't be fulfilled
+ *   |        |
+ *   +--------+
+ *     feed, need more data to serve
+ *     previous consumer request
+ *
+ *
+ *  Copyright (C) 2006-2018, ARM Limited, All Rights Reserved
```

Comment:
I followed e.g. [ecdh.h](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ecdh.h), which has the summary section prior to the copyright, too.

Reply:
Splitting the informative from the copyright section as in [ecd.h](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ecdh.h), but keeping the order for now.

Reply:
I don't know about any legal regulation regarding the copyright notice's placement. I suppose it should be clearly visible, so in the current form it may be feasible. However, it may be better to consult it with someone, who has already dealt with similar issues.